### PR TITLE
Add missing title values so they can be used by templates

### PR DESCRIPTION
**Pull request by dunxd**
I couldn't find a consistant way to put the configured title for the library into a template.

`it.fullTitle` and `it.title` are set to the same value by line 271 of JSON_renderer.php - I don't understand why.

The JavaScript variable currentData seems to contain all the variables available to doT, and JSON_renderer.php looks to be where this is created, so I added a line that exposes `$config['cops_title_default']` as `it.libraryName` which seems to be doing the trick.

Maybe I missed something else?

### CODE REVIEW

**Comment by dunxd** [comment]
I also needed the translation for recently added books in a template.  Given that other people might want to add page titles in templates, I have added all the translations that end with .title to the i18n array.

**Comment by mikespub** [comment]
Regarding fullTitle vs. title, I don't know all the history of the code base, but one trick I often use when I don't know why something was added is to:

1. find out where it's used and where it's defined (with search on Github, or git grep in code)
https://github.com/search?q=repo%3Amikespub-org%2Fseblucas-cops%20fullTitle&type=code

and then

2. see when it was added there (with Blame on Github, and return in history until it was added)
https://github.com/mikespub-org/seblucas-cops/blame/main/lib/Output/JSON_renderer.php

Or in the case of a forked repo:

1. go to the upstream repo and do the same search on commits:
https://github.com/search?q=repo%3Aseblucas%2Fcops+fullTitle&type=commits

Which leads to the same place, i.e. a commit from seblucas on May 31, 2014: "Add a fullTitle property to show the current database"

And that's what the code actually does: in case of multiple database setup, it adds the database name before the title, and shows that in the header template, but not anywhere else the title is used. Mystery solved :-)

**Comment by dunxd** [comment]
It isn't nearly ready yet.  When I feel it is, I'll submit it as a PR. I
probably won't call it bootstrap3 either :-)

It might be more of an experiment - I've spent more time trying to
understand the code and add things missing to the template engine than I
have actually designing and building the template. The handoff from PHP to
Javascript can be confusing.

Hopefully I've learnt enough and tied up enough loose ends to get it
further.  At which point I'll need to figure out whether using Bootstrap v5
instead of the one in the vendor folder is a good idea.  The old templates
break with Bootstrap v5 :-( My template uses the Bootstrap CDN to fetch
stylesheets and scripts allowing the new version, but might not be the best
for users of COPS.

**Comment by mikespub** [comment]
OK, I'll hold off on merging this then, since the bootstrap3 changes are included here, and only update JSON renderer.

And I agree with your reluctance on relying on bootstrap v5 for e-readers - I don't know enough to tell how this would affect them either. And that's also why I didn't update the jquery & js-cookie dependencies to a later version yet...

File: lib/Output/JSON_renderer.php [hunk 1]
```diff
@@ -185,12 +188,14 @@ public static function addCompleteArray($in, $request)
...
                            "pubdateTitle" => localize("pubdate.title"),
-                           "languagesTitle" => localize("language.title"),
+                           "languagesTitle" => localize("languages.title"),
+                           "languageTitle" => localize("language.title"),
                            "contentTitle" => localize("content.summary"),
                            "filterClearAll" => localize("filter.clearall"),
                            "sortorderAsc" => localize("search.sortorder.asc"),
                            "sortorderDesc" => localize("search.sortorder.desc"),
                            "customizeEmail" => localize("customize.email")],
+                           "ratingsTitle" => localize("ratings.title"),
```
Comment (mikespub) [inline lib/Output/JSON_renderer.php:198]: This looks like the ] from the previous line should be moved too, no?

Comment (dunxd) [reply lib/Output/JSON_renderer.php]: Thanks for catching that. I've fixed it and committed.

**Comment by dunxd** [comment]
With the Bootstrap2 template on my Kindle Paperwhite  I am getting the error message undefined constant "j" in $out at the top of the page, and then it is displaying a lot of the javascript code.

Bootstrap template doesn't have this issue.  Bootstrap3 (which used Bootstrap2 as a starting point) has the same issue.

Debugging the page on the Kindle is pretty much impossible - it lacks view source and doesn't allow screenshots or have a clipboard so I can't even show you what I am seeing.  A better solution would be to detect Kindle browsers and force them to use a template that works and disallow changing the template at all.

Note to self: User agent for Kindle Paperwhite is `Mozilla/5.0 (X11; U; Linux armv7l like Android;en-us) AppleWebKit/531.2+ (KHTML, like Gecko) Version/5.0 Safari/533.2+ Kindle/3/0+`

Can't tell if l in `arm7l` is a pipe or a lowercase L doesn't look the same as the 1 in X11.

**Comment by mikespub** [comment]
Hey @dunxd you might want to try and override the following config parameter in your config_local.php:

```
/*
 * Contains a list of user agent for browsers not compatible with client side rendering
 * For now : Kindle, Sony PRS-T1, Sony PRS-T2, All Cybook devices (maybe a little extreme).
 * This item is used as regular expression so "." will force server side rendering for all devices
 */
$config['cops_server_side_render'] = 'Kindle\/1\.0|Kindle\/2\.0|Kindle\/3\.0|EBRD1101|EBRD1201|cybook';
```

In your case, it should be Kindle\/3\/0 instead of Kindle\/3\.0 - see if that helps force COPS to use server side rendering (PHP) instead of client side rendering (JS).

**Comment by dunxd** [comment]
Sorry - I seem to have fat fingered the string when copying it across - it does end with `Kindle/3.0+` so should be getting server side rendering with the default `$config['cops_server_side_render']`.

- default template works fine
- bootstrap template works fine
- bootstrap2 displays errors on pages that render lists of books
- bootstrap3 displays the errors on all pages

I get the same result if I add the end of my Windows browser agent string to the end of `$config['cops_server_side_render']`.  The error messages are all truncated and scattered through the page.  Searching for what does appear doesn't return much - looks like computed code that is being evaluated.

image

I am wondering if a change I made to the code broke things.

**Comment by dunxd** [comment]
I just read the README.md for resources/dot-php which has a warning that it wasn't tested with anything other than the templates that seblucas wrote for COPS.

Feels like there are two options:
- further work on dot-php
- forcing server side rendering to use the working default template.

What do you think?

**Comment by mikespub** [comment]
Hey @dunxd if the templates work fine with dot.js and client side rendering, I'm fine trying to tweak dot-php for particular use cases to support server side rendering as well.

At least if we don't get too fancy with all kinds of dot.js features in the templates that aren't supported in dot-php at all :-)

In this particular case, I'm guessing something is wrong in main.html since that's the only place where "j" is used. This error message comes from evaluating the compiled template code in doT.php:

```
    try {
        return str_replace($replace["encode"], $replace["unsafe"], eval($func));
    } catch (\Throwable $e) {
        return $e->getMessage() . ' in ' . $func;
    }
```

So knowing what $func looks like compared to the original template, this at least gives us a place to start tracking...

Looking at view source in the browser, this particular piece seems to have some constant "j" instead of $j in the compiled code, so that's something. Now whether it can be fixed, and why it works for "default" but not "bootstrap2" that's something else :-)

```
<div class="download bottom">
  <!-- download Split button -->
  '; if ($entry["book"]["preferedData"]["length"] > 1) { $out.='<div class="btn-group">';} $out.='
      '; for ($j = 0; $j < count($entry["book"]["preferedData"] ); $j++) { $data = $entry["book"]["preferedData"]  [$j]; $out.='
          '; if (j == 0) { $out.='
              <a href="[' . $data["url"]  . '" class="btn btn-primary btn-sm" role="button"><span class="glyphicon glyphicon-download"></span> ' . $data["name"]  . '</a>
              '; if ($entry["book"]["preferedData"]["length"] > 1) { $out.='
                  <button type="button" class="btn btn-primary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                      <span class="caret"></span>
                      <span class="sr-only">Toggle Dropdown</span>
                  </button>
                  '; if ($entry["book"]["preferedData"]["length"] > 1) { $out.='<ul class="dropdown-menu">';} $out.='
              ';} $out.='
          ';} else { $out.='
              <li><a href="[' . $data["url"]  . '">' . $data["name"]  . '</a></li>
              '; if ($entry["book"]["preferedData"]["length"] = j + 1) { $out.='</ul>';} $out.='
          ';} $out.='
      ';} $out.='
  '; if ($entry["book"]["preferedData"]["length"] > 1) { $out.='</div>';} $out.='
</div>
```

I'll need to spend some time figuring out if it can be fixed, so in the meantime the best option might be to switch back to "default" template for server side rendering indeed.

**Comment by mikespub** [comment]
It looks like @SenorSmartyPants never meant bootstrap2 to be used with server side rendering, because it's using Javascript-specific methods like .length which don't apply in PHP, and array index "j" which doesn't get any special treatment in doT.php - unlike "i" which apparently got an exception to get the default and bootstrap templates working :-)

We can probably fix this particular case with an extra variable and some minor template changes, so that bootstrap2 and variants can work server side too.

But supporting Javascript array methods or making the looping index available in general is a bit more tricky to handle so that they work both client side and server side, and it's not something I want to tackle with doT.php...

**Comment by mikespub** [comment]
@dunxd I pushed some minimal changes to get bootstrap2 working with server side rendering
See https://github.com/mikespub-org/seblucas-cops/commit/a95569d37593efc912df31a5e47e726127c72d99

Key points:
- do not use Javascript methods like .length, but send any extra data needed from JSON renderer - in this case preferedCount
- if you need to refer to a loop index, only "i" will work, and only if it's at the start of the *if* condition - in this case rename other indexes + turn the *if* clause around

Quick & dirty, but it will let you play with bootstrap3 or others :-)

**Comment by dunxd** [comment]
That's great. I would never have figured that out.

My plans for Bootstrap3 are to start with Bootstrap2 but make it simpler rather than more complex. For example, I think the drop down to download different formats is less helpful than just seeing the different formats listed. 

Now I know to stay away from using JavaScript array functions. Hopefully there is no need to use any other features of JavaScript that result in issues for server side rendering.

**Comment by dunxd** [comment]
Back from holiday, and just trying out bootstrap5 on my Kindle.  Definitely hitting some issues with server side rendering.

If I use a conditional in the template (e.g. `{{? it.page==10}}` it gets converted by dot.php to `if (it"]["page==10)` and throws an error about an unexpected double quoted string "][".

I resolved this by adding spaces between the variables and the operators (e.g. ` it.page == 10`.  This doesn't cause an error in dot.js!

**Comment by mikespub** [comment]
Indeed - spaces are good :-)

I've actually been working on using Twig as template engine for COPS as an alternative for doT, and so far it seems to be working OK both server-side and client-side. That would allow a more "robust" template engine to be used for more advanced features in future templates...

For now that's in a separate [twig](https://github.com/mikespub-org/seblucas-cops/tree/twig) branch, and only used for the "twigged" template as a direct translation from bootstrap2, so I'll probably merge that in some future release

**Pull request merged**